File: PHP/TuringMachine.php
<?php
class TuringMachine {
    private $tape;
    private $head;
    private $state;

    public function __construct($input) {
        // Initialize the tape with input and extra spaces on both sides
        $this->tape = str_split("  " . $input . "  ");
        $this->head = 2; // Start at the first character of the input
        $this->state = "q0"; // Initial state
    }

    public function execute() {
        while ($this->state !== "accept" && $this->state !== "reject") {
            $currentSymbol = $this->tape[$this->head];

            switch ($this->state) {
                case "q0":
                    if ($currentSymbol === 'a') {
                        $this->tape[$this->head] = 'X';
                        $this->head++;
                        $this->state = "q1";
                    } elseif ($currentSymbol === ' ') {
                        $this->state = "accept";
                    } else {
                        $this->state = "reject";
                    }
                    break;

                case "q1":
                    if ($currentSymbol === 'a' || $currentSymbol === 'X') {
                        $this->head++;
                    } elseif ($currentSymbol === 'b') {
                        $this->tape[$this->head] = 'Y';
                        $this->head++;
                        $this->state = "q2";
                    } else {
                        $this->state = "reject";
                    }
                    break;

                case "q2":
                    if ($currentSymbol === 'b' || $currentSymbol === 'Y') {
                        $this->head++;
                    } elseif ($currentSymbol === ' ') {
                        $this->state = "accept";
                    } else {
                        $this->state = "reject";
                    }
                    break;
            }
        }

        if ($this->state === "accept") {
            echo "Input is accepted by the Turing Machine.\n";
        } else {
            echo "Input is rejected by the Turing Machine.\n";
        }
    }
}

$input = "aabb";
$tm = new TuringMachine($input);
$tm->execute();

File: Kotlin/TuringMachine.kt
class TuringMachine(input: String) {
    private val tape = input.toCharArray() + "  ".toCharArray()
    private var head = 0
    private var state = "q0"

    fun execute() {
        while (state != "accept" && state != "reject") {
            val currentSymbol = tape[head]

            when (state) {
                "q0" -> when (currentSymbol) {
                    'a' -> {
                        tape[head] = 'X'
                        head++
                        state = "q1"
                    }
                    ' ' -> state = "accept"
                    else -> state = "reject"
                }

                "q1" -> when (currentSymbol) {
                    'a', 'X' -> head++
                    'b' -> {
                        tape[head] = 'Y'
                        head++
                        state = "q2"
                    }
                    else -> state = "reject"
                }

                "q2" -> when (currentSymbol) {
                    'b', 'Y' -> head++
                    ' ' -> state = "accept"
                    else -> state = "reject"
                }
            }
        }

        if (state == "accept") {
            println("Input is accepted by the Turing Machine.")
        } else {
            println("Input is rejected by the Turing Machine.")
        }
    }
}

fun main() {
    val input = "aabb"
    val tm = TuringMachine(input)
    tm.execute()
}
